timemachine/fe/reweighting.py

```diff
@@ -0,0 +1,286 @@
+__all__ = [
+    "one_sided_exp",
+    "construct_endpoint_reweighting_estimator",
+    "construct_mixture_reweighting_estimator",
+    "interpret_as_mixture_potential",
+]
+
+from typing import Callable, Collection
+
+from jax import numpy as np
+from jax.scipy.special import logsumexp
+
+Samples = Params = Collection
+Energies = Array = np.array
```

Comment:
I'm not sure why, but this style seems to break mypy checks for me (many seemingly-unrelated errors are generated). The problem disappears when I change this to
``` python
Samples = Collection
Params = Collection
Array = np.ndarray
Energies = Array
```

Reply:
Ahh! Thanks, I hadn't been running these using mypy so far, meant these mostly as human-readable annotations.

This file introduces a bunch of mypy errors:

<details>

```text
mypy timemachine/fe/reweighting.py 
timemachine/fe/reweighting.py:15: error: Variable "timemachine.fe.reweighting.Samples" is not valid as a type
timemachine/fe/reweighting.py:15: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:15: error: Variable "timemachine.fe.reweighting.Params" is not valid as a type
timemachine/fe/reweighting.py:15: error: Variable "timemachine.fe.reweighting.Energies" is not valid as a type
timemachine/fe/reweighting.py:18: error: Variable "timemachine.fe.reweighting.Array" is not valid as a type
timemachine/fe/reweighting.py:18: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:28: error: Variable "timemachine.fe.reweighting.Array" is not valid as a type
timemachine/fe/reweighting.py:28: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:33: error: Variable "timemachine.fe.reweighting.Array" is not valid as a type
timemachine/fe/reweighting.py:33: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:43: error: Unsupported operand type for unary - (Array?)
timemachine/fe/reweighting.py:46: error: Variable "timemachine.fe.reweighting.Array" is not valid as a type
timemachine/fe/reweighting.py:46: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:101: error: Array? has no attribute "shape"
timemachine/fe/reweighting.py:103: error: Array? has no attribute "shape"
timemachine/fe/reweighting.py:106: error: Array? has no attribute "T"
timemachine/fe/reweighting.py:114: error: Variable "timemachine.fe.reweighting.Samples" is not valid as a type
timemachine/fe/reweighting.py:114: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:115: error: Variable "timemachine.fe.reweighting.Samples" is not valid as a type
timemachine/fe/reweighting.py:115: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:118: error: Variable "timemachine.fe.reweighting.Params" is not valid as a type
timemachine/fe/reweighting.py:118: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:120: error: Variable "timemachine.fe.reweighting.Params" is not valid as a type
timemachine/fe/reweighting.py:120: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:162: error: Unsupported left operand type for - (Energies?)
timemachine/fe/reweighting.py:167: error: Unsupported left operand type for - (Energies?)
timemachine/fe/reweighting.py:170: error: Variable "timemachine.fe.reweighting.Params" is not valid as a type
timemachine/fe/reweighting.py:170: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:196: error: Variable "timemachine.fe.reweighting.Samples" is not valid as a type
timemachine/fe/reweighting.py:196: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:197: error: Variable "timemachine.fe.reweighting.Array" is not valid as a type
timemachine/fe/reweighting.py:197: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:200: error: Variable "timemachine.fe.reweighting.Params" is not valid as a type
timemachine/fe/reweighting.py:200: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
timemachine/fe/reweighting.py:263: error: Unsupported operand type for unary - (Energies?)
Found 22 errors in 1 file (checked 1 source file)
```

</details>

Will see if I can fix these up

Reply:
Nice -- applying the two small suggested changes https://github.com/proteneer/timemachine/pull/654/commits/f1ad7ab50a455f89be15722645a25a2578405562 and https://github.com/proteneer/timemachine/pull/654/commits/d0b565995c76f030899c18b4b70086f0e96818d5, whittles this down to only one mypy error:

```
timemachine/fe/reweighting.py:109: error: "ndarray" has no attribute "T"
Found 1 error in 1 file (checked 1 source file)
```
on this line
 
https://github.com/proteneer/timemachine/blob/d0b565995c76f030899c18b4b70086f0e96818d5/timemachine/fe/reweighting.py#L109

which I could avoid by using `np.transpose(x)` instead of `x.T`...

err, might be a good idea to expand this out to be more self-documenting / explicit anyway (avoiding the implicit vector minus matrix = matrix rank-promotion...)

Reply:
I expanded this a bit in https://github.com/proteneer/timemachine/pull/654/commits/4dd40082698c18c34242313a30c1a0e7b6e46b22 to improve general readability

Reply:
Nice, these seem like good changes! I'm also realizing that typing the `vmap`ped code might be more trouble than it's worth -- it looks like the [jax annotation](https://github.com/google/jax/blob/c7508d1f2daa4fc6f7d74464d9d840a274c46fec/jax/_src/api.py#L1384) isn't precise enough to infer the return type of the generated function, so it seems like typing code that uses vmap might require a lot of manual annotation

Reply:
Documented these steps a little more, and retained the equivalent one-liner in a comment in https://github.com/proteneer/timemachine/pull/654/commits/9f9971ae8248f16b0b946a425935b70167db5fbb

Reply:
>I'm also realizing that typing the vmapped code might be more trouble than it's worth -- it looks like the [jax annotation](https://github.com/google/jax/blob/c7508d1f2daa4fc6f7d74464d9d840a274c46fec/jax/_src/api.py#L1384) isn't precise enough to infer the return type of the generated function, so it seems like typing code that uses vmap might require a lot of manual annotation

Hmm, yeah, the effect of jax transformations is not precisely annotated in the types (grad: Callable -> Callable, vmap: Callable -> Callable, ...).

I think it might still make sense to annotate types for (1) a function that processes "a single element" e.g. `f: Item -> float`, (2) interfaces that assume batches e.g. `batch_f: Items -> Array` -- even if static analysis might not be able to verify that applying vmap to (1) will produce a function that matches the type annotation of (2).